Create UNO game for PS

sparkychildcharlie opened the pull request:
- uses the room-game API
- should pass Travis (I swear I tested 3 times http://prntscr.com/ea22as)
- uses Fisher-Yates shuffle for the cards
- cards display will keep same dimensions and fit to most screen sizes.
(sample display of card selection screen http://prntscr.com/ea2114)

(fixed the main things talked about from one year ago in https://github.com/Zarel/Pokemon-Showdown/pull/2374)

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+}
+
+// Durstenfeld shuffle algorithm - optimized Fisher-Yates shuffle.
+function shuffle(array) {
```
Um... I don't think you need a shuffle function; ``Tools.shuffle`` is a thing

sparkychildcharlie replied:
wow.... I totally did not know that existed at all...
Serves me right for keeping sugar in for side server code haha... Array#randomize :(

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+			this.discards.unshift(this.topCard);
+		} while (this.topCard.colour === "Black");
+
+		this.sendToRoom(`|raw|The top card is <font color=${this.topCard.colour}>${this.topCard.name}</font>.`);
```
I propose we stop using ``<font>``

sparkychildcharlie replied:
I assume that the alternative is <span>?

on this note, i'll remove all <center> as well and replace it with text-align iirc.

sorry didnt realize these weren't HTML5 supported.

Zarel replied:
Yeah, `<span style="...">` is the HTML5 way to do it.

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+			return this.sendReply("UNO has been disabled for this room.");
+		},
+
+		enable: function (target, room, user) {
```
If this does get added, then a roomsettings piece to enable/disable this would probably be a good idea (to mirror tournaments, kind of)

sparkychildcharlie replied:
mhm i'll PR this if this passes.

On this note, this follows the standard set in the hangman code. @Zarel would you want something more unified to control all minigame permissions in rooms?

Zarel replied:
This works for now. In the long term we might want some sort of minigame management.

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+
+			room.game = new UNOgame(room, target);
+			room.add(`|uhtml|uno-${room.gameNumber}|<div class="broadcast-green"><p style="font-size: 14pt; text-align: center;">A new game of <strong>UNO</strong> is starting!</p><p style="font-size: 9pt; text-align: center;">Use <strong>/uno join</strong> to join the game.</p></div>`).update();
+			return this.privateModCommand("(A game of UNO was created by " + user.name + ".)");
```
Let's use template strings consistently

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+		start: function (target, room, user) {
+			if (!this.can("minigame", null, room)) return;
+			if (!room.game || room.game.gameid !== "uno" || room.game.state !== "signups") return this.errorReply("There is no UNO game in signups phase in this room.");
+			if (room.game.onStart()) this.privateModCommand("(The game of UNO was started by " + user.name + ".)");
```
Template string here, too

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+		timer: function (target, room, user) {
+			if (!this.can("minigame", null, room)) return;
+			if (!room.game || room.game.gameid !== "uno") return this.errorReply("There is no UNO game going on in this room.");
+			let amount = parseInt(target);
```
Check for ``NaN``?

Slayer95 replied:
``NaN`` is falsey, so it's covered by ``!amount``

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+
+	onStart() {
+		if (this.playerCount < 2) return false;
+		this.sendToRoom(`|uhtmlchange|uno-${this.room.gameNumber}|<div class="infobox">The game of UNO has started.</div>`);
```
``room.gameNumber`` is kind of vague. Maybe ``room.unoGameNumber`` or something?

sparkychildcharlie replied:
I'm just keeping the same format as what hangman used.

Zarel replied:
`.gameNumber` is fine.

panpawn commented:
Sidenote: ``uthml`` is not supported in battles, and from a quick peek, I don't see a check preventing the commands from being used in battles, unless I'm mistaken?

panpawn commented:
Also, screen caps of UI?

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+"use strict";
```
``'use strict';``

tic marks seems to be the standard around PS, not quote marks

panpawn commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+
+function cardHTML(card, fullsize) {
+	let surface = card.value.replace(/[^A-Z0-9\+]/g, "");
+	return `<button style="height: 140px; padding-bottom: 120px; text-align: left; font-weight: bold; border: 2px solid rgba(33 , 68 , 72 , 0.59); min-width: ${fullsize ? "80" : "30"}px; border-radius: 10px 2px 2px 3px; color: ${card.colour === "Black" ? "yellow" : "black"}; background-color: ${card.colour}" name=send value="/uno play ${card.name}">${surface}</button>`;
```
I think using ternary operators in template strings wasn't exactly a straightforward topic... still not sure on where that falls among PS standards, actually

panpawn replied:
I think it was that we should probably try to avoid it if possible, but again, not sure

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+	}
+
+	return [...basic, ...basic, // two copies of the basic stuff (total 96)
+		...[...[...Array(4)].map((v, i) => {
```
Maybe it's just me, but I'd prefer an explicit
`` new Array(4).fill(0).map(...)``
instead of
``[...Array(4)].map(...)``

, which just looks redundant in a function so spread-heavy.

Zarel replied:
I'd prefer an explicit `[0, 1, 2, 3].map(...)`

Slayer95 replied:
That works, too...

Afterwards,
``...[...[0, 1, 2, 3].map(..)``

Reduces to ``...[0, 1, 2, 3].map(..)``

Basically, ``...[...iterable]`` is equivalent to ``...iterable``. Except that the former version iterates twice.

Zarel replied:
`...[0, 1, 2, 3].map(..)` is the same as `[0, 1, 2, 3].map(..)`, right?

Zarel replied:
Oh, it's part of a larger array, nvm, I see what's going on now.

sparkychildcharlie commented:
I'll put these changes in tonight, as well as fixing a very big mistake i accidentally made - allowing players to join after the game has started :(

Zarel commented:
Battles are already roomgames; anything that is also a roomgame is automatically blocked from existing in battles.

Zarel commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,563 @@
+
+	return [...basic, ...basic, // two copies of the basic stuff (total 96)
+		...[...[...Array(4)].map((v, i) => {
+			return {colour: colours[i], value: "0", name: colours[i] + " 0"};
```
Also, `{ return {...}; }` is way too verbose of a way to write something that can just be written `({...})`

sparkychildcharlie replied:
ah thanks for teaching me! I didn't know that was possible!

Zarel replied:
Oh, yeah, the basic syntax of an arrow function is `input => output`.

So, for instance, a function that just returns its input is

`x => x`

This is short for `(x) => { return x; }`

With objects, you need parentheses around them so `({a: b})` so they'll be interpreted as objects rather than code blocks.

sparkychildcharlie commented:
- changed things to template strings
- removed center tags and replaced with HTML5 compatible stuff ``text-align: center;``
- removed font tags and with ``span style="color: #......"``
- fixed joining after the game has started
- removed the "abusive" spread operators in ``createDeck()`` and replaced them with ~~Slayer's~~ Zarel's suggestions (sorry for double commit, I didn't notice the extra comments below it)
- fixed the map syntax with parentheses

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,546 @@
+	return [...basic, ...basic, // two copies of the basic stuff (total 96)
+		...[0, 1, 2, 3].map(v => ({colour: colours[v], value: "0", name: colours[v] + " 0"})), // the 4 0s
+		...[0, 1, 2, 3].map(v => ({colour: "Black", value: "Wild", name: "Wild"})), // wild cards
+		...[0, 1, 2, 3].fill(0).map(v => ({colour: "Black", value: "+4", name: "Wild +4"})), // wild +4 cards
```
The ``.fill(0)`` call shouldn't be there.

sparkychildcharlie replied:
ah my bad, I forgot to remove that when i was switching from your method to Zarel's

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,547 @@
+			if (a.colour < b.colour) return -1;
+			if (a.value > b.value) return 1;
+			return -1;
+		})
```
This can be written as ``.sort((a, b) => (a.colour - b.colour) || (a.value - b.value))``

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,547 @@
+	hasCard(cardName) {
+		let matched = this.hand.filter(c => c.name === cardName);
+		if (!matched.length) return false;
+		return matched[0]; // returns the first instance of the matched card
```
Just ``this.hand.find(c => c.name === cardName)``

(Note that this alternative returns ``undefined`` if there is no match).

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,547 @@
+
+		// clear previous display and show new display
+		this.sendRoom("|uhtmlchange|uno-hand|");
+		this.sendRoom(`|uhtml|uno-hand|<table style="width: 100%; table-layout: fixed;">${this.game.currentPlayer === this.userid ? `<tr><td colspan=2></td><td style="border-radius: 50px 0 0 0; border: 1px solid; padding: 5px; text-align: right;">${top}</td></tr>` : ""}<tr><td colspan=2 style="border: 1px solid; padding: 5px;"><div style="overflow-x: auto; white-space: nowrap; width: 100%;">${hand}</di ...
```
Consider splitting this line into multiple lines.

Slayer95 replied:
Also this HTML code wants some client-side CSS support badly ;)

Slayer95 commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,547 @@
+				clearTimeout(room.game.timer);
+				room.game.timer = setTimeout(() => {
+					room.game.eliminate(room.game.currentPlayer);
+				}, amount * 1000);
```
There needs to be a cap though. The timeout should be less than 2^31 so that it fits in a 32-bit integer rather than overflowing.

Reject the input if ``amount * 1000 > 0x7FFFFFFF``

Slayer95 replied:
More practically, there is hardly any reason for the disqualification timeout to exceed 1000 seconds, so checking for that should be more helpful, in as much as it prevents people from trying to use milliseconds as unit.

sparkychildcharlie commented:
- used Array#find for checking if the card exists.
- split the html for hand display into several lines (had to use + to join the strings since template strings allow newlines just by pressing enter) - Do we __have to__ do it using client-side css?  I'm pretty sure that hangman and other chat plugins just list the styling stuff in the HTML as well.
- used a simpler sort ``this.hand.sort((a, b) => (a.colour > b.colour) || (a.colour === b.colour && a.value > b.value))``  Sorry Slayer, your method doesn't seem to work since it mixes between sorting by colour and value.
- limited timer to max 300 seconds (5 minutes)... PS users already complain that 150 seconds for battles is already too much...

Here's another screenshot:
![UNO](http://image.prntscr.com/image/6164b76970bc4273941c17eefb0ad9fc.png)

Slayer95 commented:
@sparkychildcharlie , maybe you wrote ``>`` in your comparison function rather than ``-``.
I'm fairly sure my method ought to work.... I also tested it :|

sparkychildcharlie commented:
when i tried a.colour ("Green") - b.colour ("Blue") = null

Slayer95 commented:
So you were comparing strings with ``>`` Sigh...

Then try this: ``.sort((a, b) => a.colour.localeCompare(b.colour) || (a.value - b.value))

sparkychildcharlie commented:
Ah i see how that works... thanks for the tip about the localeCompare! I didn't know that existed before... it's amazing :3 

(I used that for the value comparison as well, since the values are all strings to allow stuff like +2, Reverse, and Wild)

MacChaeger commented:
How does this look in dark mode?

sparkychildcharlie commented:
Dark mode viewing (uses plain default CSS from main server) - it looks good imo
![dark mode](http://image.prntscr.com/image/f740dc0464864abb801da023b32d3763.png)

MacChaeger commented:
What about the black Wild card?

On Sun, Feb 19, 2017 at 8:39 AM, Charlie Kobayashi <notifications@github.com
> wrote:

> Dark mode viewing (uses plain default CSS from main server) - it looks
> good imo
> [image: dark mode]
> <https://camo.githubusercontent.com/b8c5c67767c37d9f0d92d2ddb7a56ec4aad98359/687474703a2f2f696d6167652e70726e747363722e636f6d2f696d6167652f66373430646330343634383634616262383031646130323362333264333736332e706e67>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Zarel/Pokemon-Showdown/pull/3241#issuecomment-280923178>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AULj4K62t8YtOIBkRSJ16K0lE-eOQvXPks5reFQMgaJpZM4MEro1>
> .
>

panpawn commented:
That UI actually looks pretty good 👍

sparkychildcharlie commented:
Dark mode wild cards
![dark](http://image.prntscr.com/image/be93d929b6104ba88b40c4b92f187e7a.png)

Zarel commented:
The yellow cards look very weird. Possibly too saturated.

This design is okay, but I'm not a _huge_ fan...

KrisOnPS commented:
I could probably work out some CSS if you'd like to make it more aesthetically pleasing

sparkychildcharlie commented:
Display changes:
- done to Zarel's request/suggestions!
- hides the UNO button automatically the turn after

Drawing: 
- refactored regular drawing of a card to be an "event" of its own (onDraw)
- prevents users from drawing a card after playing a colour card (drawing can only happen in during the "play" state now)
- only drawing on your own turn will show you your hand display (you can still view your hand any time using ``/uno`` or ``/uno hand``)

UNO handling:
- handles uno parsing if the next player draws, refactored into it's own function as well (since it occurs at both the draw and the playing part)
- for 2p games, if the user does not say uno before playing their last card (Skip => [last card]), they can draw/play without having to say UNO and be penalized 2 cards (in official rules, iirc you have to beat the next player's actions, and not your own)

sparkychildcharlie commented:
(bump)

It's been updated ~7 days ago; currently waiting for further comments or for Zarel to push the big green button :3

sparkychildcharlie commented:
Bump~!

@Zarel -senpai! It's ready to be evaluated and potentially added too :3

Zarel commented:
@Slayer95?

Slayer95 commented:
Just one thing... Can I haz gradientz plz?

sparkychildcharlie commented:
What do you mean? I'm assuming you're referring to a change to the colours in the display through CSS, but I'm not exactly too sure what you'd want to see.

sparkychildcharlie commented:
- gradients to cards for Slayer-senpai
- use the button class for cards
- make pass/draw buttons less hard
- put in Zarel-senpai's recommendations for text shadows

- display updated cards after playing anything

@Zarel it's... updated! :3

Zarel commented:
The most recent screenshot you've shown me is

![image](https://cloud.githubusercontent.com/assets/551184/23668327/ba3b5914-032e-11e7-8f01-a4064f1fb56a.png)

Which I still told you to unbalance the text shadow...

[3/7/17, 8:51:37 AM] Zarel: so, something like
[3/7/17, 8:51:52 AM] Zarel: 1px 1px, -1px 1px, -1px -1px, 3px -3px
[3/7/17, 8:52:00 AM] Zarel: all 0 fuzziness and black

Zarel commented:
Also, I don't think a radial gradient is the best idea; I'd personally prefer something like a top-down gradient.

sparkychildcharlie commented:
oopsie! I thought you accidentally left that out this is top down now

![stuff](http://image.prntscr.com/image/fe868231ab10481f9ccf624ce21ac8a5.png)

Zarel commented:
That is way too much offset. Apparently you have it zoomed in a lot. Try 2px -2px.

Zarel commented:
At that zoom level, it looks like 4x might not be enough.

In which case:

`1px 0px, -1px 0px, 0px -1px, 0px 1px, 3px -3px`

sparkychildcharlie commented:
Done.  The last one was at 90% zoom on a 13-inch laptop

Zarel commented:
Nah, those are fine as-is. Just give another screenshot.

sparkychildcharlie commented:
![#team_reshiram](http://image.prntscr.com/image/a90c2e82c16d4b588c8e79a93fcbfc8a.png)
![#team_zekrom](http://image.prntscr.com/image/b8cbce9659834bb3a9b23a0071195908.png)

panpawn commented:
I think that text shadow is a bit much

acz13 commented:
yeah, UNO cards don't have that much shadow IIRC

Zarel commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,576 @@
+	"Black": "rgba(0, 0, 0, 1), rgba(0, 0, 0, 0.65)",
+};
+
+const text_shadow = "text-shadow: 1px 0px black, -1px 0px black, 0px -1px black, 0px 1px black, 3px -3px black;";
```
I said `2px -2px`...

sparkychildcharlie replied:
"At that zoom level, it looks like 4x might not be enough.

In which case:

1px 0px, -1px 0px, 0px -1px, 0px 1px, 3px -3px"

oh i thought you wanted this.....

Zarel replied:
Sorry, that one was a typo. The line above that said `2px -2px` was what I meant. :s

Zarel replied:
But the others are right

sparkychildcharlie replied:
ok updated, sorry for the wait, choir and other "educational" things happened first

sparkychildcharlie commented:
![Shadow ((and sonic hehehe))](http://image.prntscr.com/image/9bbedc844488410a8c7696d7ce96a028.png) 
@Zarel - I assume you're probably waiting for the screenshot :D

Zarel commented:
I think this looks pretty good, but what happened to the gradient?

sparkychildcharlie commented:
I believe it's still there.  The colour is slightly lighter on the bottom.  Just a touch and not too much to make it seem awkward

Zarel commented:
Oh, that's very subtle. I can see it in the green now but the others have basically no contrast. Yeah, you might want to tune it up a bit.

sparkychildcharlie commented:
![img](http://image.prntscr.com/image/35a1c25e29a6441884f64fa96d7b1f57.png)

Zarel commented:
The yellow gradient is hard to notice. Maybe it should be darker on top? Maybe look at official Uno cards for inspiration; their yellow cards definitely don't look so whitish.

sparkychildcharlie commented:
![yellow-chan](http://image.prntscr.com/image/eeed8f80046a45bc9c25ce659249051b.png)

Changed the colours to use more of 2 colours as opposed to simply transparency. The yellow card starts at a slightly more orange hue before going down to a "pure" yellow at the bottom.

Zarel commented:
Actually looks pretty good, although the yellow card might be too orangey.

sparkychildcharlie commented:
@Zarel 
![http://image.prntscr.com/image/782fc76eef194b90ae2972d7713d7e68.png](http://image.prntscr.com/image/782fc76eef194b90ae2972d7713d7e68.png)

Making it only use Yellow (and yellow mixed with black for the darker hue).  There isnt much differnece otherwise it wouldn't appear to be yellow anymore

panpawn commented:
I think that looks pretty good.

Zarel commented:
I'd say less orangey and not no orange at all; this gradient is no longer even noticeable... :/

Zarel commented:
Also your blue tints purple for some reason...

sparkychildcharlie commented:
That's weird, the blue is rgb(0,0,255) which should be max blue and nothing else

sparkychildcharlie commented:
![Haydn - The Creation - Let There be Light!](http://image.prntscr.com/image/814c28419cf048ddb8193502891c52b7.png)
![Mozart - Die Zauberflöte - Queen of the Night](http://image.prntscr.com/image/49b61a31e5e947f0a87aa9880f4f60e3.png)

@Zarel - is this ok?  The gradients are noticeable much more easily for the yellows the the blues are more "luminous" and "blue"

Zarel commented:
I mean, it's good enough at this point. I'd probably want to tweak the colors myself, but we can do that after merge. Does anyone not want to merge this?

Zarel commented on chat-plugins/uno.js:
```diff
@@ -0,0 +1,576 @@
+		do {
+			this.topCard = this.drawCard(1)[0];
+			this.discards.unshift(this.topCard);
+		} while (this.topCard.colour === "Black");
```
Um, maybe use `.color` since afaik the rest of the codebase is in US English? I won't insist though.

acz13 commented:
looks fine

sparkychildcharlie commented:
bump @Zarel

Zarel commented:
...I keep accidentally clicking rebase-merge instead of squash-merge... :/
